_fcmaescpp/gcldeoptimizer.cpp

```diff
@@ -145,208 +149,219 @@ class GclDeOptimizer {
+			if (distance < minDist) {
+				index = j;
+				minDist = distance;
+			}
+		}
+		return index;
+	}
+
+	void doOptimize() {
+		int Gr = 2;
+		int gen_stuck = 0;
+		vector<vec> sp;
+		vector<InfoStore> storage;
+
+		int maxIter = maxEvaluations / popsize + 1;
+		double previous_best = DBL_MAX;
+		double CR = CR0;
+		double F = F0;
+
+		// -------------------- Generation Loop --------------------------------
+
+		for (iterations = 1; fitfun->getEvaluations() < maxEvaluations;
+				iterations++) {
+			mat nextX = mat(popX);
+			vec nextY = vec(popY);
+			vec popCR = vec(popsize);
+			vec popF = vec(popsize);
+			vec xb = popX.col(0);
+			for (int p = 0; p < popsize; p++) {
+				vec xi = popX.col(p);
+				int r1, r2, r3;
+				do {
+					r1 = rndInt(popsize);
+				} while (r1 == p);
+				do {
+					r2 = rndInt(int(popsize * pbest));
+				} while (r2 == p || r2 == r1);
+				do {
+					r3 = rndInt(popsize + sp.size());
+				} while (r3 == p || r3 == r2 || r3 == r1);
+				int jr = rndInt(dim);
+				//Produce the CR and F
+				float mu = 1
+						- sqrt(float(iterations / maxIter))
+								* exp(float(-gen_stuck / iterations));
+				if (storage.empty() || iterations < Gr) {
+					CR = normreal(rs, 0.95, 0.01);
+					F = normreal(rs, mu, 1);
+					if (F < 0 || F > 1)
+						F = rnd01();
+				} else {
+					int i = findNear(popX.col(p), storage);
+					CR = storage[i].CR;
+					F = storage[i].F;
+				}
+				popCR[p] = CR;
+				popF[p] = F;
+				vec ui = vec(xi);
+				for (int j = 0; j < dim; j++) {
+					if (j == jr || rnd01() < CR) {
+						if (r3 < popsize)
+							ui[j] = popX(j, r1)
+									+ F * (popX(j, r2) - popX(j, r3));
+						else
+							ui[j] = popX(j, r1)
+									+ F * (popX(j, r2) - sp[r3 - popsize][j]);
+						if (ui[j] > 1 || ui[j] < 0)
+							ui[j] = rnd01();
+					}
+				}
+				nextX.col(p) = ui;
+			}
+			vec fitness = vec(popsize);
+			fitfun->values(nextX, popsize, fitness);
+			for (int p = 0; p < popsize; p++) {
 //            	fitness[p] = fitfun->value(nextX.col(p));
-                if (!isfinite(fitness[p]))
-                	fitness[p] = DBL_MAX;
-            }
-            for (int p = 0; p < popsize; p++) {
-                double y = fitness[p];
-                if (y < popY[p]) {
+				if (!isfinite(fitness[p]))
+					fitness[p] = DBL_MAX;
+			}
+			for (int p = 0; p < popsize; p++) {
+				double y = fitness[p];
+				if (y < popY[p]) {
 					nextY[p] = y;
-                    if (y < bestValue) {
+					if (y < bestValue) {
 						bestValue = y;
 						bestX = nextX.col(p);
 						if (isfinite(stopfitness) && bestValue < stopfitness) {
 							stop = 1;
 							return;
 						}
 					}
-                 } else // no improvement
-                	 nextX.col(p) = popX.col(p);
-            }
-    		for (int p = 0; p < popsize; p++) {
-    			if (nextY[p] < popY[p]) { // improvement
-    				if (sp.size() < popsize)
-    				    sp.push_back(nextX.col(p));
-    				else
-    					sp[rndInt(popsize)] = nextX.col(p);

```

Comment:
Also here

Reply:
FIxed at both locations and pushed the changes to the branch.